File: MeshDebugger/Lattice.cpp
#include "Lattice.h"

#include <string>

namespace slicer {

std::string toString(LatticeType type) {
  switch (type) {
    case LatticeType::DIAMOND:
      return "diamond";
    case LatticeType::CUBE:
      return "cube";
    case LatticeType::GYROID:
      return "gyroid";
    case LatticeType::FLUORITE:
      return "fluorite";
    case LatticeType::OCTET:
      return "octet";
    default:
      return "unknown";
  }
}

LatticeType ParseLatticeType(const std::string& s) {
  LatticeType t = LatticeType::DIAMOND;
  if (s == "diamond") {
    t = LatticeType::DIAMOND;
  } else if (s == "cube") {
    t = LatticeType::CUBE;
  } else if (s == "gyroid") {
    t = LatticeType::GYROID;
  } else if (s == "fluorite") {
    t = LatticeType::FLUORITE;
  } else if (s == "octet") {
    t = LatticeType::OCTET;
  }
  return t;
}

static const float INIT_CELL_DIST = 10000.0f;
/// point line segment distance
double PtLineSegDist(const Vec3f& p, const Vec3f& l0, const Vec3f& l1) {
  Vec3f l = l1 - l0;
  Vec3f l0p = p - l0;
  if (l0p.dot(l) <= 0.0f || l.norm() < 1e-6) {
    return l0p.norm();
  }
  Vec3f l1p = p - l1;
  if (l1p.dot(l) >= 0.0f) {
    return l1p.norm();
  }
  Vec3f c = l.cross(l0p);
  return c.norm() / l.norm();
}

/// draw distance field for a beam on top of existing grid.
///  very inefficient. called only during init.
void DrawBeamDist(const Vec3f& p0, const Vec3f& p1, Array3Df& grid, const Vec3f& voxelSize) {
  Vec3u s = grid.GetSize();
  for (unsigned x = 0; x < s[0]; x++) {
    for (unsigned y = 0; y < s[1]; y++) {
      for (unsigned z = 0; z < s[2]; z++) {
        Vec3f p(x, y, z);
        p = p * voxelSize;
        double d = PtLineSegDist(p, p0, p1);
        if (d < grid(x, y, z)) {
          grid(x, y, z) = d;
        }
      }
    }
  }
}

void DrawBeams(unsigned numBeams, float beams[][6], Array3Df& grid, const Vec3f& voxelSize,
               const Vec3f& cellSize) {
  for (unsigned b = 0; b < numBeams; b++) {
    Vec3f p0(beams[b][0], beams[b][1], beams[b][2]);
    Vec3f p1(beams[b][3], beams[b][4], beams[b][5]);
    p0 = cellSize * p0;
    p1 = cellSize * p1;
    DrawBeamDist(p0, p1, grid, voxelSize);
  }
}

Array3Df MakeCubeCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  const unsigned NUM_BEAMS = 3;
  float beams[NUM_BEAMS][6] = {
      {0, 0.5, 0.5, 1, 0.5, 0.5}, {0.5, 0, 0.5, 0.5, 1, 0.5}, {0.5, 0.5, 0, 0.5, 0.5f, 1}};
  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

Array3Df MakeDiamondCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  const unsigned NUM_BEAMS = 28;
  // 16 beams within the cell and 12 beams for beams outside of the cell.
  float beams[NUM_BEAMS][6] = {{0.25, 0.25, 0.25, 0, 0, 0},      {0.25, 0.25, 0.25, .5f, .5f, 0},
                               {0.25, 0.25, 0.25, .5f, 0, .5f},  {0.25, 0.25, 0.25, 0, .5f, .5f},

                               {0.75, 0.75, 0.25, 1, 1, 0},      {0.75, 0.75, 0.25, .5f, .5f, 0},
                               {0.75, 0.75, 0.25, 1, .5f, .5f},  {0.75, 0.75, 0.25, .5f, 1, .5f},

                               {0.25, 0.75, 0.75, .5f, .5f, 1},  {0.25, 0.75, 0.75, 0, 1, 1},
                               {0.25, 0.75, 0.75, 0, .5f, .5f},  {0.25, 0.75, 0.75, .5f, 1, 0.5},

                               {0.75, 0.25, 0.75, 1, 0, 1},      {0.75, 0.25, 0.75, .5f, 0, .5f},
                               {0.75, 0.25, 0.75, 1, .5f, 0.5f}, {0.75, 0.25, 0.75, 0.5f, .5f, 1},

                               {1, 0, 0, 1.25, 0.25, 0.25},      {1, 0, 0, 0.75, -0.25, 0.25},
                               {1, 0, 0, 0.75, 0.75, -0.25},

                               {0, 1, 0, -0.25, 0.75, 0.25},     {0, 1, 0, 0.25, 1.25, 0.25},
                               {0, 1, 0, 0.75, 0.75, -0.25},

                               {0, 0, 1, -0.25, 0.25, 0.75},     {0, 0, 1, 0.25, -0.25, 0.75},
                               {0, 0, 1, 0.25, 0.25, 1.25},

                               {1, 1, 1, 1.25, 0.75, 0.75},      {1, 1, 1, 0.75, 1.25, 0.75},
                               {1, 1, 1, 0.75, 0.75, 1.25}};

  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

Array3Df MakeGyroidCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  float twoPi = 6.2831853f;
  float maxVal = 1.3f;
  Vec3u size = grid.GetSize();
  float distScale = 0.32f * cellSize[0] / std::abs(maxVal);
  for (unsigned z = 0; z < size[2]; z++) {
    float fz = float(z) / size[2] * twoPi;
    for (unsigned y = 0; y < size[1]; y++) {
      float fy = float(y) / size[1] * twoPi;
      for (unsigned x = 0; x < size[0]; x++) {
        float fx = float(x) / size[0] * twoPi;
        float distVal =
            std::sin(fx) * std::cos(fy) + std::sin(fy) * std::cos(fz) + std::sin(fz) * std::cos(fx);
        grid(x, y, z) = distScale * std::abs(distVal);
      }
    }
  }
  return grid;
}

Array3Df MakeFluoriteCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  const unsigned NUM_BEAMS = 32;
  float beams[NUM_BEAMS][6];
  const unsigned NUM_F_VERTS = 8;
  const unsigned NUM_CA_VERTS = 14;
  Vec3f FVerts[NUM_F_VERTS] = {{0.25f, 0.25f, 0.25f}, {0.75f, 0.25f, 0.25f}, {0.25f, 0.75f, 0.25f},
                               {0.75f, 0.75f, 0.25f}, {0.25f, 0.25f, 0.75f}, {0.75f, 0.25f, 0.75f},
                               {0.25f, 0.75f, 0.75f}, {0.75f, 0.75f, 0.75f}};
  Vec3f CaVerts[NUM_CA_VERTS] = {{0, 0, 0},       {1, 0, 0},       {0, 1, 0},       {1, 1, 0},
                                 {0, 0, 1},       {1, 0, 1},       {0, 1, 1},       {1, 1, 1},
                                 {0.5f, 0.5f, 0}, {0.5f, 0.5f, 1}, {0.5f, 0, 0.5f}, {0.5f, 1, 0.5f},
                                 {0, 0.5f, 0.5f}, {1, 0.5f, 0.5f}};
  unsigned beamIndex = 0;
  float distThresh = 0.44f;
  for (unsigned i = 0; i < NUM_F_VERTS; i++) {
    for (unsigned j = 0; j < NUM_CA_VERTS; j++) {
      float dist = (FVerts[i] - CaVerts[j]).norm();
      if (dist < distThresh && beamIndex < NUM_BEAMS) {
        for (unsigned d = 0; d < 3; d++) {
          beams[beamIndex][d] = FVerts[i][d];
          beams[beamIndex][d + 3] = CaVerts[j][d];
        }
        beamIndex++;
      }
    }
  }
  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

Array3Df MakeOctetCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  const unsigned NUM_BEAMS = 20;
  float beams[NUM_BEAMS][6] = {{0, 0, 0, 1, 1, 0},
                               {1, 0, 0, 0, 1, 0},
                               {0, 0, 1, 1, 1, 1},
                               {1, 0, 1, 0, 1, 1},
                               {0, 0, 0, 1, 0, 1},
                               {1, 0, 0, 0, 0, 1},
                               {0, 1, 0, 1, 1, 1},
                               {1, 1, 0, 0, 1, 1},
                               {0, 0, 0, 0, 1, 1},
                               {0, 0, 1, 0, 1, 0},
                               {1, 0, 0, 1, 1, 1},
                               {1, 0, 1, 1, 1, 0},

                               {0.5f, 0.5f, 0, 0.5f, 0, 0.5f},
                               {0.5f, 0.5f, 0, 0.5f, 1, 0.5f},
                               {0.5f, 0.5f, 0, 0, 0.5f, 0.5f},
                               {0.5f, 0.5f, 0, 1, 0.5f, 0.5f},
                               {0.5f, 0.5f, 1, 0.5f, 0, 0.5f},
                               {0.5f, 0.5f, 1, 0.5f, 1, 0.5f},
                               {0.5f, 0.5f, 1, 0, 0.5f, 0.5f},
                               {0.5f, 0.5f, 1, 1, 0.5f, 0.5f}};

  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

Array3Df MakeShearXCell(Vec3f voxelSize, Vec3f cellSize, Vec3u gridSize) {
  Array3Df grid;
  grid.Allocate(gridSize, INIT_CELL_DIST);
  const unsigned NUM_BEAMS = 12 + 2 * 4;
  float beams[NUM_BEAMS][6] = {
      {0, 0, 0, 1, 0, 0},   {0, 0, 1, 1, 0, 1},   {0, 1, 1, 1, 1, 1},
      {0, 1, 0, 1, 1, 0},   {0, 0, 0, 0, 1, 0},   {0, 0, 1, 0, 1, 1},
      {1, 0, 0, 1, 1, 0},   {1, 0, 1, 1, 1, 1},   {0, 0, 0, 0.2, 0, 1},
      {0, 1, 0, 0.2, 1, 1}, {0.8, 0, 0, 1, 0, 1}, {0.8, 1, 0, 1, 1, 1},
      {0, 0, 0, 0.2, 1, 1}, {0.2, 0, 1, 0, 1, 0}, {0.8, 0, 0, 1, 1, 1},
      {0.8, 1, 0, 1, 0, 1}, {0, 0, 0, 1, 1, 0},   {1, 0, 0, 0, 1, 0},
      {0, 0, 1, 1, 1, 1},   {0, 1, 1, 1, 0, 1}};

  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

}  // namespace slicer

File: Load3dm/src/Lattice.ts
import { Array3D } from "./Array3D.js";
import { Vec3f } from "./Vec3f.js";

/// point line segment distance
function PtLineSegDist(p, l0, l1) {
  const l = l1.subtract(l0);
  const l0p = p.subtract(l0);
  if (l0p.dot(l) <= 0.0 || l.norm() < 1e-6) {
    return l0p.norm();
  }
  const l1p = p.subtract(l1);
  if (l1p.dot(l) >= 0.0) {
    return l1p.norm();
  }
  const c = l.cross(l0p);
  return c.norm() / l.norm();
}

/// draw distance field for a beam on top of existing grid.
///  very inefficient. called only during init.
export function DrawBeamDist(p0, p1, grid, voxelSize) {
  const s = grid.size;
  const p = new Vec3f();
  for (let x = 0; x < s[0]; x++) {
    for (let y = 0; y < s[1]; y++) {
      for (let z = 0; z < s[2]; z++) {
        p.Set(x, y, z);
        const pmm = p.Mult(voxelSize);
        const d = PtLineSegDist(pmm, p0, p1);
        if (d < grid.Get(x, y, z)) {
          grid.Set(x, y, z, d);
        }
      }
    }
  }
}

export function DrawBeams(numBeams, beams, grid, voxelSize, cellSize) {
  for (let b = 0; b < numBeams; b++) {
    const p0 = new Vec3f(beams[b][0], beams[b][1], beams[b][2]);
    const p1 = new Vec3f(beams[b][3], beams[b][4], beams[b][5]);
    const p0mm = p0.Mult(cellSize);
    const p1mm = p1.Mult(cellSize);
    DrawBeamDist(p0mm, p1mm, grid, voxelSize);
  }
}

export function MakeDiamondCell(voxelSize, cellSize, grid: Array3D) {
  grid.data.fill(10000);
  const NUM_BEAMS = 28;
  // 16 beams within the cell and 12 beams for beams outside of the cell.
  const beams = [
    [0.25, 0.25, 0.25, 0, 0, 0],
    [0.25, 0.25, 0.25, 0.5, 0.5, 0],
    [0.25, 0.25, 0.25, 0.5, 0, 0.5],
    [0.25, 0.25, 0.25, 0, 0.5, 0.5],

    [0.75, 0.75, 0.25, 1, 1, 0],
    [0.75, 0.75, 0.25, 0.5, 0.5, 0],
    [0.75, 0.75, 0.25, 1, 0.5, 0.5],
    [0.75, 0.75, 0.25, 0.5, 1, 0.5],

    [0.25, 0.75, 0.75, 0.5, 0.5, 1],
    [0.25, 0.75, 0.75, 0, 1, 1],
    [0.25, 0.75, 0.75, 0, 0.5, 0.5],
    [0.25, 0.75, 0.75, 0.5, 1, 0.5],

    [0.75, 0.25, 0.75, 1, 0, 1],
    [0.75, 0.25, 0.75, 0.5, 0, 0.5],
    [0.75, 0.25, 0.75, 1, 0.5, 0.5],
    [0.75, 0.25, 0.75, 0.5, 0.5, 1],

    [1, 0, 0, 1.25, 0.25, 0.25],
    [1, 0, 0, 0.75, -0.25, 0.25],
    [1, 0, 0, 0.75, 0.25, -0.25],

    [0, 1, 0, -0.25, 0.75, 0.25],
    [0, 1, 0, 0.25, 1.25, 0.25],
    [0, 1, 0, 0.25, 0.75, -0.25],

    [0, 0, 1, -0.25, 0.25, 0.75],
    [0, 0, 1, 0.25, -0.25, 0.75],
    [0, 0, 1, 0.25, 0.25, 1.25],

    [1, 1, 1, 1.25, 0.75, 0.75],
    [1, 1, 1, 0.75, 1.25, 0.75],
    [1, 1, 1, 0.75, 0.75, 1.25],
  ];

  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

export function MakeGyroidCell(_voxelSize, cellSize, grid) {
  grid.data.fill(10000);
  const twoPi = 6.2831853;
  const maxVal = 1.3;
  const size = grid.size;
  const distScale = (0.32 / maxVal) * cellSize.x;

  const stepx = (1.0 / (size[0] - 1)) * twoPi;
  const stepy = (1.0 / (size[1] - 1)) * twoPi;
  const stepz = (1.0 / (size[2] - 1)) * twoPi;
  for (let z = 0; z < size[2]; z++) {
    const fz = z * stepz;
    for (let y = 0; y < size[1]; y++) {
      const fy = y * stepy;
      for (let x = 0; x < size[0]; x++) {
        const fx = x * stepx;
        const distVal =
          Math.sin(fx) * Math.cos(fy) +
          Math.sin(fy) * Math.cos(fz) +
          Math.sin(fz) * Math.cos(fx);
        grid.Set(x, y, z, distScale * Math.abs(distVal));
      }
    }
  }
}

export function MakeFluoriteCell(voxelSize, cellSize, grid) {
  grid.data.fill(10000);
  const NUM_BEAMS = 32;
  const beams = Array.from({ length: 32 }, () => Array(6));
  const NUM_F_VERTS = 8;
  const NUM_CA_VERTS = 14;
  const FVerts = [
    [0.25, 0.25, 0.25],
    [0.75, 0.25, 0.25],
    [0.25, 0.75, 0.25],
    [0.75, 0.75, 0.25],
    [0.25, 0.25, 0.75],
    [0.75, 0.25, 0.75],
    [0.25, 0.75, 0.75],
    [0.75, 0.75, 0.75],
  ];
  const CaVerts = [
    [0, 0, 0],
    [1, 0, 0],
    [0, 1, 0],
    [1, 1, 0],
    [0, 0, 1],
    [1, 0, 1],
    [0, 1, 1],
    [1, 1, 1],
    [0.5, 0.5, 0],
    [0.5, 0.5, 1],
    [0.5, 0, 0.5],
    [0.5, 1, 0.5],
    [0, 0.5, 0.5],
    [1, 0.5, 0.5],
  ];
  let beamIndex = 0;
  const distThresh = 0.44;
  const va = new Vec3f();
  const vb = new Vec3f();
  for (let i = 0; i < NUM_F_VERTS; i++) {
    for (let j = 0; j < NUM_CA_VERTS; j++) {
      va.Set(FVerts[i][0], FVerts[i][1], FVerts[i][2]);
      vb.Set(CaVerts[j][0], CaVerts[j][1], CaVerts[j][2]);
      const dist = va.subtract(vb).norm();
      if (dist < distThresh && beamIndex < NUM_BEAMS) {
        for (let d = 0; d < 3; d++) {
          beams[beamIndex][d] = FVerts[i][d];
          beams[beamIndex][d + 3] = CaVerts[j][d];
        }
        beamIndex++;
      }
    }
  }
  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}

export function MakeOctetCell(voxelSize, cellSize, grid) {
  grid.data.fill(10000);
  const NUM_BEAMS = 20;
  const beams = [
    [0, 0, 0, 1, 1, 0],
    [1, 0, 0, 0, 1, 0],
    [0, 0, 1, 1, 1, 1],
    [1, 0, 1, 0, 1, 1],
    [0, 0, 0, 1, 0, 1],
    [1, 0, 0, 0, 0, 1],
    [0, 1, 0, 1, 1, 1],
    [1, 1, 0, 0, 1, 1],
    [0, 0, 0, 0, 1, 1],
    [0, 0, 1, 0, 1, 0],
    [1, 0, 0, 1, 1, 1],
    [1, 0, 1, 1, 1, 0],

    [0.5, 0.5, 0, 0.5, 0, 0.5],
    [0.5, 0.5, 0, 0.5, 1, 0.5],
    [0.5, 0.5, 0, 0, 0.5, 0.5],
    [0.5, 0.5, 0, 1, 0.5, 0.5],
    [0.5, 0.5, 1, 0.5, 0, 0.5],
    [0.5, 0.5, 1, 0.5, 1, 0.5],
    [0.5, 0.5, 1, 0, 0.5, 0.5],
    [0.5, 0.5, 1, 1, 0.5, 0.5],
  ];

  DrawBeams(NUM_BEAMS, beams, grid, voxelSize, cellSize);
  return grid;
}
